lib/Crypto/crypto_c.h

```diff
@@ -0,0 +1,302 @@
+};
+
+enum Sailfish_Crypto_Key_EncryptionPadding {
+    Sailfish_Crypto_Key_EncryptionPaddingUnknown    = 0,
+    Sailfish_Crypto_Key_EncryptionPaddingNone       = 1,
+    Sailfish_Crypto_Key_EncryptionPaddingPkcs7      = 2,
+    Sailfish_Crypto_Key_EncryptionPaddingRsaOaep    = 4,
+    Sailfish_Crypto_Key_EncryptionPaddingRsaOaepMgf1= 8,
+    Sailfish_Crypto_Key_EncryptionPaddingRsaPkcs1   = 16,
+    Sailfish_Crypto_Key_EncryptionPaddingAnsiX923   = 32
+};
+
+enum Sailfish_Crypto_Key_SignaturePadding {
+    Sailfish_Crypto_Key_SignaturePaddingUnknown     = 0,
+    Sailfish_Crypto_Key_SignaturePaddingNone        = 1,
+    Sailfish_Crypto_Key_SignaturePaddingRsaPss      = 2,
+    Sailfish_Crypto_Key_SignaturePaddingRsaPkcs1    = Sailfish_Crypto_Key_EncryptionPaddingRsaPkcs1,
+    Sailfish_Crypto_Key_SignaturePaddingAnsiX923    = Sailfish_Crypto_Key_EncryptionPaddingAnsiX923
+};
+
+enum Sailfish_Crypto_Key_Digest {
+    Sailfish_Crypto_Key_DigestUnknown       = 0,
+    Sailfish_Crypto_Key_DigestSha1          = 1,
+    Sailfish_Crypto_Key_DigestSha256        = 2,
+    Sailfish_Crypto_Key_DigestSha384        = 4,
+    Sailfish_Crypto_Key_DigestSha512        = 8
+};
+
+enum Sailfish_Crypto_Key_Operation {
+    Sailfish_Crypto_Key_OperationUnknown    = 0,
+    Sailfish_Crypto_Key_Sign                = 1,
+    Sailfish_Crypto_Key_Verify              = 2,
+    Sailfish_Crypto_Key_Encrypt             = 4,
+    Sailfish_Crypto_Key_Decrypt             = 8
+};
+
+struct Sailfish_Crypto_Key {
+    struct Sailfish_Crypto_Key_Identifier *identifier;
+
+    enum Sailfish_Crypto_Key_Origin origin;
+    enum Sailfish_Crypto_Key_Algorithm algorithm;
+    int blockModes;
+    int encryptionPaddings;
+    int signaturePaddings;
+    int digests;
+    int operations;
+
+    unsigned char *secretKey;
+    size_t secretKeySize;
+    unsigned char *publicKey;
+    size_t publicKeySize;
+    unsigned char *privateKey;
+    size_t privateKeySize;
+
+    time_t validityStart;
+    time_t validityEnd;
+
+    struct Sailfish_Crypto_Key_CustomParameter *customParameters;
+    struct Sailfish_Crypto_Key_FilterDatum *filterData;
+};
+
+struct Sailfish_Crypto_Key_Identifier*
+Sailfish_Crypto_Key_Identifier_new(
+        const char *name,
+        const char *collectionName);
+
+void Sailfish_Crypto_Key_Identifier_delete(
+        struct Sailfish_Crypto_Key_Identifier *ident);
+
+struct Sailfish_Crypto_Key_FilterDatum*
+Sailfish_Crypto_Key_FilterDatum_new(
+        const char *field,
+        const char *value);
+
+void Sailfish_Crypto_Key_FilterDatum_delete(
+        struct Sailfish_Crypto_Key_FilterDatum *filter);
+
+struct Sailfish_Crypto_Key_CustomParameter*
+Sailfish_Crypto_Key_CustomParameter_new(
+        const unsigned char *parameter,
+        size_t parameterSize);
+
+void Sailfish_Crypto_Key_CustomParameter_delete(
+        struct Sailfish_Crypto_Key_CustomParameter *param);
+
+struct Sailfish_Crypto_Key*
+Sailfish_Crypto_Key_new(
```

Comment:
Would we want the C user to create new keys by him or the only way to obtain keys are using the cryptoManager functions? Should this new() function be moved out of the public header?

Reply:
Client applications can create new keys themselves, or they can create new template keys and ask a crypto plugin to generate a key based on the template, or they can create a reference key (which contains just a valid identifier).  Some might argue that the first use-case needn't be supported, but certainly the last two do need to be supported.  So I think we need this method public.